File: terrabutler/env.py
from terrabutler.settings import get_settings
from terrabutler.tf import terraform_init_all_sites
from click import confirm
from colorama import Fore
import boto3
import os
import subprocess

# Values from Settings
org = get_settings()["general"]["organization"]
default_env_name = get_settings()["environments"]["default"]["name"]
backend_dir = os.path.realpath(get_settings()["locations"]["backend_dir"])
environment_file = os.path.realpath(get_settings()
                                    ["locations"]["environment_file"])
inception_dir = os.path.realpath(get_settings()["locations"]["inception_dir"])
templates_dir = os.path.realpath(get_settings()["locations"]["templates_dir"])
variables_dir = os.path.realpath(get_settings()["locations"]["variables_dir"])


def create_env(env, confirmation, temporary, apply, s3):
    from terrabutler.settings import write_settings
    from terrabutler.tf import terraform_apply_all_sites
    from terrabutler.variables import generate_var_files
    available_envs = get_available_envs(s3)

    if env in available_envs:
        print(Fore.YELLOW + "The environment you are trying to create already"
              " exists.\n\nNo changes were made.")
        exit(1)
    elif confirmation or confirm(f"Do you really want to create '{env}'" +
                                 " environment?", default=False):
        try:
            subprocess.run(args=['terraform', 'workspace', 'new', env],
                           cwd=inception_dir, stdout=subprocess.DEVNULL,
                           stderr=subprocess.DEVNULL, check=True)
        except subprocess.CalledProcessError:
            print(Fore.RED + "There was an error while creating the new"
                  " environment.")
            exit(1)
        if temporary:  # Generate only for temporary environments
            generate_var_files(env)
        else:  # When permanent environment add to the list
            config_file = get_settings()  # Original config file
            envs = config_file["environments"]["permanent"]
            envs.append(env)
            write_settings(config_file)
        if apply and temporary:
            terraform_apply_all_sites()
        reload_direnv()
        print(Fore.GREEN + f"Created and switched to the environment '{env}'!")
    else:
        print(Fore.RED + "Creation cancelled.")
        exit(1)


def delete_env(env, confirmation, destroy, s3):
    from terrabutler.tf import terraform_destroy_all_sites
    available_envs = get_available_envs(s3)
    current_env = get_current_env()
    permanent_environments = get_settings()["environments"]["permanent"]

    if env not in available_envs:
        print(Fore.RED + "The environment you are trying to delete does not"
              " exist.\n\nNo changes were made.")
        exit(1)
    elif env == current_env:
        print(Fore.RED + "The environment you are trying to delete is your"
              " active environment.\n\nPlease switch to another workspace "
              "and try again.")
        exit(1)
    elif env in permanent_environments:
        print(Fore.RED + "The environment you are trying to delete is a" +
              " permanent environment and can not be deleted.\n\nNo changes" +
              " were made.")
        exit(1)
    elif confirmation or confirm(f"Do you really want to delete '{env}'" +
                                 " environment?", default=False):
        if destroy and not is_protected_env(env):
            terraform_destroy_all_sites()  # Destroy all sites
        for file in os.listdir(variables_dir):
            if file.startswith(f"{org}-{env}"):
                os.remove(os.path.join(variables_dir, file))
        try:
            subprocess.run(args=['terraform', 'workspace', 'delete', env],
                           cwd=inception_dir, stdout=subprocess.DEVNULL,
                           stderr=subprocess.DEVNULL, check=True)
        except subprocess.CalledProcessError:
            print(Fore.RED + f"There was an error while deleting the '{env}'"
                  "environment.")
            exit(1)
        print(Fore.GREEN + f"The environment '{env}' was deleted!")
    else:
        print(Fore.RED + "Deletion cancelled.")
        exit(1)


def get_current_env():
    with open(environment_file, 'r') as f:
        return f.read()


def set_current_env(env, s3):
    current_env = get_current_env()
    available_envs = get_available_envs(s3)

    if env == current_env:
        print(Fore.YELLOW + "The environment you selected is the current one."
              "\n\nNo changes were made.")
        exit(0)
    elif env not in available_envs:
        print(Fore.RED + f"The environment '{env}' does not exist.\n\nYou can"
              " create this environment with the 'new' command.")
        exit(1)
    else:
        try:
            with open(environment_file, "w") as f:
                f.write(env)
        except FileNotFoundError:
            print(Fore.RED + "The file that manages the environments could not"
                  " be found.")
            exit(1)

        reload_direnv()
        terraform_init_all_sites()
        print("\n\n" + Fore.GREEN + f"Switched to environment '{env}'.")


def get_available_envs(s3):

    envs = []

    # Get Environments by accessing S3
    if s3:
        dev_env = boto3.session.Session(profile_name=f"{org}"
                                        f"-{default_env_name}")
        s3 = dev_env.resource("s3")
        bucket = s3.Bucket(f"{org}-{default_env_name}-site-inception-tfstate")

        envs = []

        for folder in bucket.objects.filter(Prefix="env:/", Delimiter=""):
            folder_name = folder.key.split('/')[1].strip()
            envs.append(folder_name)

        return envs

    # Get Environments by accessing the .terraform/environment file
    directory = inception_dir
    subprocess.run(args=["terraform", "init", "-reconfigure",
                         "-backend-config",
                         f"{backend_dir}/{org}-{default_env_name}-"
                         "inception.tfvars"],
                   cwd=directory,
                   stdout=subprocess.DEVNULL,
                   stderr=subprocess.DEVNULL)
    process = subprocess.run(args=['terraform', 'workspace', 'list'],
                             cwd=directory,
                             stdout=subprocess.PIPE,
                             stderr=subprocess.DEVNULL,
                             universal_newlines=False)
    workspaces = process.stdout.splitlines()
    if len(workspaces) > 1:  # Only run if we have already done a init
        workspaces.pop()  # Remove null entry
        workspaces.pop(0)  # Pop default workspace
    for workspace in workspaces:
        envs.append(workspace.decode(
            "utf-8").replace(" ", "").replace("*", ""))
    return envs


def reload_direnv():
    try:
        subprocess.run(args=['direnv', 'reload'])
    except subprocess.CalledProcessError as e:
        print(e.output)


def is_protected_env(env):
    protected_envs = get_settings()["environments"]["permanent"]
    if env in protected_envs:
        return True
    return False

File: terrabutler/env.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"slices"
	"strings"

	"go.uber.org/zap"
)

func confirmation_menu(question string) bool {
	fmt.Printf("%s [y/N]: ", question)
	var input string
	_, err := fmt.Scanln(&input)
	if err != nil {
		logger.Error("Fail to read input.", zap.Error(err))
	}
	input = strings.ToLower(strings.TrimSpace(input))
	if input == "y" || input == "yes" {
		return true
	}
	return false
}

func get_current_env() string {

	//Open site environment file
	env, err := os.ReadFile(paths["environment"])
	if err != nil {
		logger.Error("An error has occurred:", zap.Error(err))
		os.Exit(1)
	}
	logger.Debug("Current environment is " + string(env))
	return string(env)
}

func is_protected_env(env string) bool {

	if slices.Contains(settings.Strings("environments.permanent"), env) {
		return true
	}
	return false
}

func get_available_envs() []string {

	//List of env
	var envs []string

	directory := paths["inception"]
	org := settings.String("general.organization")
	default_env_name := settings.String("environments.default.name")

	os.Chdir(directory)

	cmd := exec.Command("terraform", "init", "-backend-config", paths["backend"]+"/"+org+"-"+default_env_name+"-inception.tfvars")
	err := cmd.Run()
	if err != nil {
		logger.Error("There was an error from terraform init for "+org+"-"+default_env_name+" environment.", zap.Error(err))
		os.Exit(1)
	}

	cmd = exec.Command("terraform", "workspace", "list")
	workspaces, err := cmd.Output()
	if err != nil {
		logger.Error("There was an error from terraform workspace for "+org+"-"+default_env_name+" environment.", zap.Error(err))
		os.Exit(1)
	}
	//After getting the output string, its needed to
	//Remove the new line
	workspaces_trim := strings.ReplaceAll(string(workspaces), "\n", "")
	//Remove '*' and substitute with " " to still have the same number of " "
	workspaces_trim = strings.ReplaceAll(workspaces_trim, "*", " ")
	//Split the workspaces into a slice flag (Normally between each one exists 2 spaces)
	workspaces_split := strings.Split(workspaces_trim, "  ")
	if len(workspaces_split) > 1 {
		//Get the size of workspace
		size := len(workspaces_split)
		//Remove the first two members of the workspace (" " and default)
		workspaces_split = workspaces_split[2:size]
	}
	envs = append(envs, workspaces_split...)

	return envs
}

func set_current_env(env string, init bool) {

	current_env := get_current_env()
	available_envs := get_available_envs()

	//Check if env is the current in use
	if env == current_env {
		logger.Warn("The environment you selected is the current one.")
		logger.Warn("No changes were made.")
		os.Exit(0)
	}
	//Check if env exists
	if !slices.Contains(available_envs, env) {
		logger.Error("The environment '" + env + "' does not exist.")
		logger.Error("You can create this environment with the 'new' command.")
		os.Exit(1)
	}
	pre_hook := settings.String("hooks.pre_env_select")

	if pre_hook != "" {
		command := strings.Split(pre_hook, " ")
		cmd := exec.Command(command[0], command[1:len(command)]...)
		cmd.Env = []string{
			"TERRABUTLER_OLD_ENV=" + current_env,
			"TERRABUTLER_NEW_ENV=" + env}
		err := cmd.Run()
		if err != nil {
			logger.Error("pre_env_select hook failed:", zap.Error(err))
			os.Exit(1)
		}

	}
	//Get the pre_hook -> "pre_env_select", if exists
	//Run pre_hook [CHANGE ENVIRONMENT: "TERRABUTLER_OLD_ENV": current_env, "TERRABUTLER_NEW_ENV": env]
	//Show error if it occurs

	//Try opening the file in path environments and writing the new env
	//Show error if it fails
	f, err := os.Create(paths["environment"])
	if err != nil {
		logger.Error(fmt.Sprint("An error has occurred opening the file: ", err))
		os.Exit(1)
	}
	l, err := f.Write([]byte(env))
	if l == 0 && err != nil {
		logger.Error(fmt.Sprint("An error has occurred writing to the file: ", err))
		f.Close()
		os.Exit(1)
	}
	err = f.Close()

	//If init true, run terraform_init_all_sites
	if init {
		tf_init_all_sites()
	}

	//Get the post_hook -> "post_env_select", if exists
	//Run post_hook [CHANGE ENVIRONMENT: "TERRABUTLER_OLD_ENV": current_env, "TERRABUTLER_NEW_ENV": env]
	//Show error if it occurs
	post_hook := settings.String("hooks.post_env_select")

	if post_hook != "" {
		command := strings.Split(post_hook, " ")
		cmd := exec.Command(command[0], command[1:len(command)]...)
		cmd.Env = []string{
			"TERRABUTLER_OLD_ENV=" + current_env,
			"TERRABUTLER_NEW_ENV=" + env}
		err := cmd.Run()
		if err != nil {
			logger.Error("post_env_select hook failed:", zap.Error(err))
			os.Exit(1)
		}

	}

	//Show successfully message at the end
	logger.Info("Switched to environment '" + env + "'.")

}

func delete_env(env string, confirmation bool, destroy bool) {

	org := settings.String("general.organization")

	//Check if env does exist
	if !slices.Contains(get_available_envs(), env) {
		logger.Error("The environment you are trying to delete does not exist.")
		logger.Error("No changes were made.")
		os.Exit(1)
	}
	//Check if the env is the current in use
	if env == get_current_env() {
		logger.Error("The environment you are trying to delete is your active environment.")
		logger.Error("Please switch to another workspace and try again.")
		os.Exit(1)
	}
	//Check if the env is permanent / use --> is_protected_env
	if is_protected_env(env) {
		logger.Error("The environment you are trying to delete is a permanent environment and can not be deleted.")
		logger.Error("No changes were made.")
		os.Exit(1)
	}

	//Confirmation menu, if yes
	if confirmation || confirmation_menu("Do you really want to delete '"+env+"' environment?") {

		//If destroy and is not a permanent env ^ already checked above, run tf_destroy_all_sites(env)
		if destroy {
			tf_destroy_all_sites()
		}
		//For each file in variables, remove/delete it
		files, _ := os.ReadDir(paths["variables"])
		for _, file := range files {
			if !file.IsDir() {
				fmt.Println(file.Name())
				if strings.Contains(file.Name(), org+"-"+env) {
					os.Remove(paths["variables"] + "/" + file.Name())
				}
			}
		}

		//Run the terraform workspace delete env [Path Inception] (with output and check)
		os.Chdir(paths["inception"])
		cmd := exec.Command("terraform", "workspace", "delete", env)
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
		err := cmd.Run()

		//Show a error message if process not executed correctly
		if err != nil {
			logger.Error("here was an error while deleting the new environment:", zap.Error(err))
			os.Exit(1)
		}

		//In the end show a successfully executed message
		logger.Info("The environment '" + env + "' was deleted!")
	} else {
		logger.Error("Deletion cancelled.")
		os.Exit(1)
	}

}

// TODO: variables.go for the temporary environments can use the templates
func create_env(env string, confirmation bool, temporary bool, apply bool) {

	//Check if env already exists
	if slices.Contains(get_available_envs(), env) {
		logger.Warn("The environment you are trying to create already exists.")
		logger.Warn("No changes were made.")
		os.Exit(1)
	}

	//Make a confirmation menu
	//If confirmation is true execute the function
	if confirmation || confirmation_menu("Do you really want to create '"+env+"' environment?") {
		//Run Terraform workspace new env [path inception] (with output and check)
		os.Chdir(paths["inception"])
		cmd := exec.Command("terraform", "workspace", "new", env)
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
		err := cmd.Run()
		if err != nil {
			logger.Error("here was an error while creating the new environment:", zap.Error(err))
			os.Exit(1)
		}

		//If temporary is true, generate the var files for the env
		if temporary {
			//generate_var_files(env)
		} else {
			//Else is a permanent environment for the list
			// Get the config file, append the new env to the config file and write the new config file
			envs := settings.Strings("environments.permanent")
			envs = append(envs, env)
			err := settings.Set("environments.permanent", envs)
			if err != nil {
				logger.Error("Error adding the new environment to the config file.", zap.Error(err))
			}
			write_settings(settings)

		}
		//If temporary and apply are true, terraform_apply_all_sites
		//Should OR apply?
		if temporary && apply {
			tf_apply_all_sites()
		}

		//In the end show a successfully executed message
		logger.Info("Created and switched to the environment '" + env + "'!")
	} else {
		logger.Error("Creation cancelled.")
		os.Exit(1)
	}

}
